# Panthera Leo

In [1]:
import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)
import Bio
print("Biopython v" + Bio.__version__)
from timeit import default_timer as timer
import copy
import math
import sys
from gtfparse import read_gtf





Biopython v1.78


In [2]:
from Bio import SeqIO
count = 0
sequences = [] # Here we are setting up an array to save our sequences for the next step

for seq_record in SeqIO.parse("./Panthera_leo/Panthera_leo.PanLeo1.0.dna.primary_assembly.A1.fa", "fasta"):
    if (count < 6):
        sequences.append(seq_record)
        print("Id: " + seq_record.id + " \t " + "Length: " + str("{:,d}".format(len(seq_record))) )
        print(repr(seq_record.seq) + "\n")
        count = count + 1



Id: A1 	 Length: 150,556,863
Seq('ACCCTGACCCTAACCATTCTGGTCCTGAAGAAGTCTCAGCCCCAGCCCTGCGTC...TTA')



In [3]:
sequences

[SeqRecord(seq=Seq('ACCCTGACCCTAACCATTCTGGTCCTGAAGAAGTCTCAGCCCCAGCCCTGCGTC...TTA'), id='A1', name='A1', description='A1 dna:primary_assembly primary_assembly:PanLeo1.0:A1:1:150556863:1 REF', dbxrefs=[])]

In [4]:
gtf = read_gtf('./Panthera_leo/Panthera_leo.PanLeo1.0.104.gtf')

INFO:root:Extracted GTF attributes: ['gene_id', 'gene_version', 'gene_name', 'gene_source', 'gene_biotype', 'transcript_id', 'transcript_version', 'transcript_name', 'transcript_source', 'transcript_biotype', 'exon_number', 'exon_id', 'exon_version', 'protein_id', 'protein_version', 'projection_parent_transcript']


In [6]:
gtf.head(10)

seqname   source      feature     start       end  score strand frame  \
0      C2  ensembl         gene  53348037  53389547    NaN      +     0   
1      C2  ensembl   transcript  53348037  53389547    NaN      +     0   
2      C2  ensembl         exon  53348037  53348319    NaN      +     0   
3      C2  ensembl          CDS  53348293  53348319    NaN      +     0   
4      C2  ensembl  start_codon  53348293  53348295    NaN      +     0   
5      C2  ensembl         exon  53352917  53352984    NaN      +     0   
6      C2  ensembl          CDS  53352917  53352984    NaN      +     0   
7      C2  ensembl         exon  53354920  53355030    NaN      +     0   
8      C2  ensembl          CDS  53354920  53355030    NaN      +     1   
9      C2  ensembl         exon  53357705  53357950    NaN      +     0   

              gene_id gene_version  ... transcript_version transcript_name  \
0  ENSPLOG00000004688            1  ...                                      
1  ENSPLOG00000004688            1  ...                  1       CWC22-201   
2  ENSPLOG00000004688            1  ...                  1       CWC22-201   
3  ENSPLOG00000004688            1  ...                  1       CWC22-201   
4  ENSPLOG00000004688            1  ...                  1       CWC22-201   
5  ENSPLOG00000004688            1  ...                  1       CWC22-201   
6  ENSPLOG00000004688            1  ...                  1       CWC22-201   
7  ENSPLOG00000004688            1  ...                  1       CWC22-201   
8  ENSPLOG00000004688            1  ...                  1       CWC22-201   
9  ENSPLOG00000004688            1  ...                  1       CWC22-201   

  transcript_source transcript_biotype exon_number             exon_id  \
0                                                                        
1           ensembl     protein_coding                                   
2           ensembl     protein_coding           1  ENSPLOE00000043121   
3           ensembl     protein_coding           1                       
4           ensembl     protein_coding           1                       
5           ensembl     protein_coding           2  ENSPLOE00000043128   
6           ensembl     protein_coding           2                       
7           ensembl     protein_coding           3  ENSPLOE00000043129   
8           ensembl     protein_coding           3                       
9           ensembl     protein_coding           4  ENSPLOE00000043130   

  exon_version          protein_id protein_version  \
0                                                    
1                                                    
2            1                                       
3               ENSPLOP00000006420               1   
4                                                    
5            1                                       
6               ENSPLOP00000006420               1   
7            1                                       
8               ENSPLOP00000006420               1   
9            1                                       

  projection_parent_transcript  
0                               
1                               
2                               
3                               
4                               
5                               
6                               
7                               
8                               
9                               

[10 rows x 24 columns]

In [5]:
gtf['seqname'].unique()

array(['C2', 'B2', 'A3', 'D1', 'A1', 'B4', 'C1', 'B1', 'B3', 'D2', 'E1',
       'D4', 'D3', 'A2', 'X', 'F2', 'E2', 'E3', 'F1', 'VOCV01000391.1',
       'VOCV01000222.1', 'VOCV01000149.1', 'VOCV01000384.1',
       'VOCV01000250.1', 'VOCV01000867.1', 'VOCV01000390.1',
       'VOCV01000310.1', 'VOCV01000249.1', 'VOCV01000381.1',
       'VOCV01000260.1', 'VOCV01000737.1', 'VOCV01000387.1',
       'VOCV01000414.1', 'VOCV01000188.1', 'VOCV01000388.1',
       'VOCV01000305.1', 'VOCV01000383.1', 'VOCV01000262.1',
       'VOCV01000220.1', 'VOCV01000263.1', 'VOCV01000412.1',
       'VOCV01000266.1', 'VOCV01000337.1', 'VOCV01000378.1',
       'VOCV01000269.1', 'VOCV01000385.1', 'VOCV01000306.1',
       'VOCV01000311.1', 'VOCV01000357.1', 'VOCV01000338.1',
       'VOCV01000307.1', 'VOCV01000379.1', 'VOCV01000595.1',
       'VOCV01000259.1', 'VOCV01000413.1', 'VOCV01000270.1',
       'VOCV01000499.1', 'VOCV01000333.1', 'VOCV01000540.1',
       'VOCV01000472.1', 'VOCV01000382.1', 'VOCV01000897.1',
 

In [6]:
gtfi = gtf.drop(['source','gene_source','transcript_id','transcript_version','transcript_name','gene_name','projection_parent_transcript','gene_version', 'gene_id','transcript_source','score'],axis = 1)

In [7]:
gtf.loc[gtf['seqname'] == 'A1'].iloc[7]

seqname                                         A1
source                                     ensembl
feature                                start_codon
start                                    131160467
end                                      131160469
score                                          NaN
strand                                           -
frame                                            0
gene_id                         ENSPLOG00000000004
gene_version                                     1
gene_name                                    WNT9A
gene_source                                ensembl
gene_biotype                        protein_coding
transcript_id                   ENSPLOT00000000004
transcript_version                               1
transcript_name                          WNT9A-201
transcript_source                          ensembl
transcript_biotype                  protein_coding
exon_number                                      1
exon_id                        

In [24]:
gtfi_a1 = gtfi.loc[gtfi['seqname'] == 'A1']

In [25]:
gtfi_a1.iloc[:50]

seqname         feature      start        end strand frame  \
230630      A1            gene   42950264   42950370      -     0   
230631      A1      transcript   42950264   42950370      -     0   
230632      A1            exon   42950264   42950370      -     0   
230633      A1            gene  131140027  131160469      -     0   
230634      A1      transcript  131140027  131160469      -     0   
230635      A1            exon  131160375  131160469      -     0   
230636      A1             CDS  131160375  131160469      -     0   
230637      A1     start_codon  131160467  131160469      -     0   
230638      A1            exon  131143468  131143724      -     0   
230639      A1             CDS  131143468  131143724      -     1   
230640      A1            exon  131142411  131142673      -     0   
230641      A1             CDS  131142411  131142673      -     2   
230642      A1            exon  131140027  131140509      -     0   
230643      A1             CDS  131140030  131140509      -     0   
230644      A1      stop_codon  131140027  131140029      -     0   
230645      A1            gene  117005116  117024112      +     0   
230646      A1      transcript  117005116  117024112      +     0   
230647      A1            exon  117005116  117005240      +     0   
230648      A1             CDS  117005116  117005240      +     0   
230649      A1     start_codon  117005116  117005118      +     0   
230650      A1            exon  117010220  117010353      +     0   
230651      A1             CDS  117010220  117010353      +     1   
230652      A1            exon  117022705  117022841      +     0   
230653      A1             CDS  117022705  117022841      +     2   
230654      A1            exon  117023972  117024112      +     0   
230655      A1             CDS  117023972  117024109      +     0   
230656      A1      stop_codon  117024110  117024112      +     0   
230657      A1            gene  117044857  117156435      -     0   
230658      A1      transcript  117044857  117156435      -     0   
230659      A1            exon  117155229  117156435      -     0   
230660      A1             CDS  117155229  117155606      -     0   
230661      A1     start_codon  117155604  117155606      -     0   
230662      A1            exon  117103844  117104113      -     0   
230663      A1             CDS  117103844  117104113      -     0   
230664      A1            exon  117100555  117100650      -     0   
230665      A1             CDS  117100555  117100650      -     0   
230666      A1            exon  117092369  117092692      -     0   
230667      A1             CDS  117092369  117092692      -     0   
230668      A1            exon  117060623  117060714      -     0   
230669      A1             CDS  117060623  117060714      -     0   
230670      A1            exon  117059204  117059374      -     0   
230671      A1             CDS  117059204  117059374      -     1   
230672      A1            exon  117051669  117051904      -     0   
230673      A1             CDS  117051669  117051904      -     1   
230674      A1            exon  117047808  117048169      -     0   
230675      A1             CDS  117047808  117048169      -     2   
230676      A1            exon  117044857  117045132      -     0   
230677      A1             CDS  117044860  117045132      -     0   
230678      A1      stop_codon  117044857  117044859      -     0   
230679      A1  five_prime_utr  117155607  117156435      -     0   

          gene_biotype transcript_biotype exon_number             exon_id  \
230630           snRNA                                                      
230631           snRNA              snRNA                                   
230632           snRNA              snRNA           1  ENSPLOE00000000016   
230633  protein_coding                                                      
230634  protein_coding     protein_coding                                   
230635  protein_coding   

In [27]:
gtfi_a1.iloc[50:100]

seqname      feature      start        end strand frame  \
230680      A1         gene  123714259  123733104      +     0   
230681      A1   transcript  123714259  123733104      +     0   
230682      A1         exon  123714259  123714387      +     0   
230683      A1          CDS  123714259  123714387      +     0   
230684      A1  start_codon  123714259  123714261      +     0   
230685      A1         exon  123721834  123722039      +     0   
230686      A1          CDS  123721834  123722039      +     0   
230687      A1         exon  123722706  123722806      +     0   
230688      A1          CDS  123722706  123722806      +     1   
230689      A1         exon  123729384  123729502      +     0   
230690      A1          CDS  123729384  123729502      +     2   
230691      A1         exon  123733021  123733098      +     0   
230692      A1          CDS  123733021  123733098      +     0   
230693      A1         exon  123733102  123733104      +     0   
230694      A1   stop_codon  123733102  123733104      +     0   
230695      A1         gene  117167400  117168415      -     0   
230696      A1   transcript  117167400  117168415      -     0   
230697      A1         exon  117168231  117168415      -     0   
230698      A1          CDS  117168231  117168415      -     0   
230699      A1  start_codon  117168413  117168415      -     0   
230700      A1         exon  117167400  117167601      -     0   
230701      A1          CDS  117167403  117167601      -     1   
230702      A1   stop_codon  117167400  117167402      -     0   
230703      A1         gene  117205405  117217992      +     0   
230704      A1   transcript  117205405  117217992      +     0   
230705      A1         exon  117205405  117205740      +     0   
230706      A1          CDS  117205405  117205740      +     0   
230707      A1  start_codon  117205405  117205407      +     0   
230708      A1         exon  117212269  117212496      +     0   
230709      A1          CDS  117212269  117212496      +     0   
230710      A1         exon  117217810  117217992      +     0   
230711      A1          CDS  117217810  117217989      +     0   
230712      A1   stop_codon  117217990  117217992      +     0   
230713      A1         gene  123756994  123762269      -     0   
230714      A1   transcript  123756994  123762269      -     0   
230715      A1         exon  123762173  123762269      -     0   
230716      A1          CDS  123762173  123762269      -     0   
230717      A1  start_codon  123762267  123762269      -     0   
230718      A1         exon  123762110  123762171      -     0   
230719      A1          CDS  123762110  123762171      -     2   
230720      A1         exon  123761758  123761896      -     0   
230721      A1          CDS  123761758  123761896      -     0   
230722      A1         exon  123759751  123759876      -     0   
230723      A1          CDS  123759751  123759876      -     2   
230724      A1         exon  123758907  123759049      -     0   
230725      A1          CDS  123758907  123759049      -     2   
230726      A1         exon  123756994  123757515      -     0   
230727      A1          CDS  123756997  123757515      -     0   
230728      A1   stop_codon  123756994  123756996      -     0   
230729      A1         gene  131187813  131228577      +     0   

          gene_biotype transcript_biotype exon_number             exon_id  \
230680  protein_coding                                                      
230681  protein_coding     protein_coding                                   
230682  protein_coding     protein_coding           1  ENSPLOE00000000039   
230683  protein_coding     protein_coding           1                       
230684  protein_coding     protein_coding           1                       
230685  protein_coding     protein_coding           2  ENSPLOE00000000040   
230686  protein_coding     protein_coding           2                       
230687  protein_coding  

We are only interested in the transicription reagion. So loop through and find all the regios.

Example of a Reverse gene with negative strand

transicription = 131140027	131160469

start codon = 131160467	 131160469	 We include both of the number so it is always (i-1,j)

Exon = 131160375	131160469

exon = 131143468	131143724

stop codon = 131140027	131140029

In [12]:
sequences[0].seq.complement()[131160466:131160469]

Seq('GTA')

In [14]:
sequences[0].seq.complement()[131160374:131160469]

Seq('GGCTTCATCCGCCGGCTCCCCGCGTCGCGCCGCTCGTCGTCGCAGTCGCGCTTC...GTA')

In [17]:
sequences[0].seq.complement()[131160372:131160469]

Seq('TGGGCTTCATCCGCCGGCTCCCCGCGTCGCGCCGCTCGTCGTCGCAGTCGCGCT...GTA')

Here the last TG is a donor site

Next Exon

In [18]:
sequences[0].seq.complement()[131143467:131143724]

Seq('GAGCGAACTCGTTCGACCGGGCCATCGCCGGGAGGTTCCATGTCAAGGTCGCGA...TCG')

Lets look at the acceptor region. The AG is the acceptor

In [19]:
sequences[0].seq.complement()[131143467:131143726]

Seq('GAGCGAACTCGTTCGACCGGGCCATCGCCGGGAGGTTCCATGTCAAGGTCGCGA...GGA')

In [20]:
sequences[0].seq.complement()[131140026:131140029]

Seq('AGT')

Lets look at a straight gene on the positive strand

Gene    117005116	117024112

Exon 117005116	117005240
 
Exon 117010220	117010353

In [22]:
sequences[0].seq[117005115:117005240]

Seq('ATGAACTGGTTTGGAAATTACTTCTCTAGATTTCCTGAGGATGTTTCCTTACTC...GAG')

In [23]:
sequences[0].seq[117005115:117005242]

Seq('ATGAACTGGTTTGGAAATTACTTCTCTAGATTTCCTGAGGATGTTTCCTTACTC...GGT')

Notice the last GT in this sequence. It is the donor site.

In [5]:
gtf['chrom'].unique()

array(['chr2L', 'chr2R', 'chr3L', 'chr3R', 'chr4', 'chrUn_CP007080v1',
       'chrUn_CP007081v1', 'chrUn_CP007097v1', 'chrUn_DS483755v1',
       'chrUn_DS483780v1', 'chrUn_DS483783v1', 'chrUn_DS483801v1',
       'chrUn_DS483878v1', 'chrUn_DS484006v1', 'chrUn_DS484105v1',
       'chrUn_DS484134v1', 'chrUn_DS484375v1', 'chrUn_DS484521v1',
       'chrUn_DS484581v1', 'chrUn_DS484779v1', 'chrUn_DS485083v1',
       'chrUn_DS485107v1', 'chrUn_DS485610v1', 'chrX',
       'chrX_DS483995v1_random', 'chrY', 'chrY_CP007109v1_random',
       'chrY_CP007113v1_random', 'chrY_CP007116v1_random',
       'chrY_DS483742v1_random', 'chrY_DS483778v1_random',
       'chrY_DS483875v1_random', 'chrY_DS483931v1_random',
       'chrY_DS484390v1_random', 'chrY_DS484643v1_random',
       'chrY_DS485048v1_random', 'chrY_DS485159v1_random'], dtype=object)

Lets take a look at all of the tags that we will need now.

EL1, EL1, EL3 - exons with low GC count

EH1, EH2, EH3 - exons with high GC count

ES1, ES2, ES3 - for single strand genes

I1, I2, I3 - For introns less than 100 in length

Id1, Id2, Id3 - for introns before D = 100

ID1, ID2, ID3 - for introns after D = 100

ASS - for 2 acceptor points

DSS - for 2 Donor points



In [29]:
def check_single_exon(i, Y):
    y = Y.iloc[i]
    count_exon = 0
    while y['feature'] != 'gene':
        if y['feature'] == 'exon':
            count_exon += 1
        i += 1
        y = Y.iloc[i]
    return count_exon == 1

def code_single_exon(start, end, is_reverse):
    y = []
    num = int((end-start+1)/3)
    numr = (end-start+1)%3
    if is_reverse:
        y += ['_E1', '_E2', '_E3'] * num
        if numr == 1:
            y+=['_E1']
        elif numr == 2:
            y+=['_E1', '_E2']
    else:
        y += ['E1', 'E2', 'E3'] * num
        if numr == 1:
            y+=['E1']
        elif numr == 2:
            y+=['E1', 'E2']
    return y


def code_exon(start, end, is_reverse, X, last_exon):
    y = []
    g = X.count('G')
    c = X.count('C')
    a = X.count('A')
    t = X.count('T')
    GC = True if (g+c)/(a+t+g+c) > 0.5 else False
    if is_reverse:
        if GC:
            num = int((end - start + 1)/3)
            numr = (end-start+1)%3
            if last_exon == '3':
                y = ['_EH1','_EH2','_EH3'] * num
            elif last_exon == '2':
                y = ['_EH3','_EH1','_EH2'] * num
            elif last_exon == '1':
                y = ['_EH2','_EH3','_EH1'] * num
            if y[-1] == '_EH1':
                if numr == 2:
                    y += ['_EH2', '_EH3']
                elif numr == 1:
                    y += ['_EH2']
            elif y[-1] == '_EH2':
                if numr == 2:
                    y += ['_EH3', '_EH1']
                elif numr == 1:
                    y += ['_EH3']
            elif y[-1] == '_EH3':
                if numr == 2:
                    y += ['_EH1', '_EH2']
                elif numr == 1:
                    y += ['_EH1']
        else:
            num = int((start - end + 1)/3)
            numr = (start-end+1)%3
            if last_exon == '3':
                y = ['_EL1','_EL2','_EL3'] * num
            elif last_exon == '2':
                y = ['_EL3','_EL1','_EL2'] * num
            elif last_exon == '1':
                y = ['_EL2','_EL3','_EL1'] * num
            if y[-1] == '_EL1':
                if numr == 2:
                    y += ['_EL2', '_EL3']
                elif numr == 1:
                    y += ['_EL2']
            elif y[-1] == '_EL2':
                if numr == 2:
                    y += ['_EH3', '_EH1']
                elif numr == 1:
                    y += ['_EH3']
            elif y[-1] == '_EH3':
                if numr == 2:
                    y += ['_EH1', '_EH2']
                elif numr == 1:
                    y += ['_EH1']
        
    else:
        if GC:
            num = int((start - end + 1)/3)
            numr = (start-end+1)%3
            if last_exon == '3':
                y = ['EH1','EH2','EH3'] * num
            elif last_exon == '2':
                y = ['EH3','EH1','EH2'] * num
            elif last_exon == '1':
                y = ['EH2','EH3','EH1'] * num
            if y[-1] == 'EH1':
                if numr == 2:
                    y += ['EH2', 'EH3']
                elif numr == 1:
                    y += ['EH2']
            elif y[-1] == 'EH2':
                if numr == 2:
                    y += ['EH3', 'EH1']
                elif numr == 1:
                    y += ['EH3']
            elif y[-1] == 'EH3':
                if numr == 2:
                    y += ['EH1', 'EH2']
                elif numr == 1:
                    y += ['EH1']
        else:
            num = int((start - end + 1)/3)
            numr = (start-end+1)%3
            if last_exon == '3':
                y = ['EL1','EL2','EL3'] * num
            elif last_exon == '2':
                y = ['EL3','EL1','EL2'] * num
            elif last_exon == '1':
                y = ['EL2','EL3','EL1'] * num
            if y[-1] == 'EL1':
                if numr == 2:
                    y += ['EL2', 'EL3']
                elif numr == 1:
                    y += ['EL2']
            elif y[-1] == 'EL2':
                if numr == 2:
                    y += ['EH3', 'EH1']
                elif numr == 1:
                    y += ['EH3']
            elif y[-1] == 'EH3':
                if numr == 2:
                    y += ['EH1', 'EH2']
                elif numr == 1:
                    y += ['EH1']
        last_exon = y[-1][-1]
        return last_exon, y

def code_intron(start, end, is_reverse):
    y = []
    end = end - 1
    length = end - start - 4
    if is_reverse:
        if length > 100:
            y = ['_Id1','_Id2','_Id3']*33
            y += ['_Id1']
            num = int((length-100)/3)
            numr = (length - 100)%3
            y += ['_ID2', '_ID3', '_ID1'] * num
            if numr == 1:
                y+=['_ID2']
            elif numr == 2:
                y+=['_ID2', '_ID3']

        else:
            num = int((length-100)/3)
            numr = (length - 100)%3
            y += ['_I1', '_I2', '_I3'] * num
            if numr == 1:
                y+=['_I1']
            elif numr == 2:
                y+=['_I1', '_I2']
        y = ['_ASS', '_ASS'] + y + ['_DSS','_DSS']
    else:
        if length > 100:
            y = ['Id1','Id2','Id3']*33
            y += ['Id1']
            num = int((length-100)/3)
            numr = (length - 100)%3
            y += ['ID2', 'ID3', 'ID1'] * num
            if numr == 1:
                y+=['ID2']
            elif numr == 2:
                y+=['ID2', 'ID3']

        else:
            num = int((length-100)/3)
            numr = (length - 100)%3
            y += ['I1', 'I2', 'I3'] * num
            if numr == 1:
                y+=['I1']
            elif numr == 2:
                y+=['I1', 'I2']
        y = ['ASS', 'ASS'] + y + ['DSS','DSS']

    return y

        

def code_gene(i , X , Y, is_reverse,strand, single_exon):
    gene_x = []
    gene_y = []
    y = Y.iloc[i]
    if y['feature'] == 'transcription':
        if strand:
            gene_x = np.array([rec for rec in str(X.upper())],dtype='U')[start-1:end]
        else:
            gene_x = np.array([rec for rec in str(X.complement().upper())],dtype='U')[start-1:end]
    else:
        print('error')
    first_exon = True
    last_exon = '3'
    end_exon = 0
    while y['feature'] != 'gene':
        if y['feature'] == 'exon' and first_exon:
            start = int(y['start'])
            end = int(y['end'])
            end_exon = end
            if single_exon:
                gene_y.append(code_single_exon(start, end, is_reverse))
            else:
                last_exon, lst = code_exon(start, end, is_reverse ,X, last_exon)
                gene_y.append(lst)
            first_exon = False
            
        elif y['feature'] == 'exon' and not first_exon:
            start = int(y['start'])
            end = int(y['end'])
            lst = code_intron(end_exon, start, is_reverse) # the index of end_exon and start is not included. 
            gene_y.append(lst)
            last_exon, lst = code_exon(start, end, is_reverse ,X, last_exon)
            gene_y.append(lst)

        i += 1

        y = Y.iloc[i]

    return i, gene_x.tolist(), gene_y

def check_gene(i, Y):
    y = Y.iloc[i]
    start_codon = 0
    stop_codon = 0
    strand = y['strand'] == '+'
    while y['feature'] != 'gene':
        if y['feature'] == 'start_codon':
            start_codon = int(y['start'])
        elif y['feature'] == 'stop_codon':
            stop_codon = int(y['start'])

        i+=1
        y = Y.iloc[i]
    return start_codon > stop_codon, strand # returns True if the strand is positive and doesnt need ot be complemented

def create_dataset(X, Y):
    i = 0
    df = pd.DataFrame(columns = ['x','y'])
    while i < len(Y):
        y = Y.iloc[i]
        if y['gene_biotype'] == 'protein_coding' and y['feature'] == 'gene':
            is_reverse, strand = check_gene(i, Y)
            single_exon = check_single_exon(i,Y)
            i,__x, __y = code_gene(i+1,X, Y, is_reverse, strand, single_exon)
            datapoint = pd.Series({'x': __x, 'y':__y})
            df = df.append(datapoint)

In [ ]:
create_dataset(sequences[0].seq, gtfi_a1)

In [6]:
def code_Exon_Intron(coding_row, x, y):
    ch[coding_row['txStart']:coding_row['txEnd']] = 'T'
    Estart = np.array(coding_row['exonStarts'].split(',')[:-1]).astype(int)
    Eend = np.array(coding_row['exonEnds'].split(',')[:-1]).astype(int)
    y_star = y[Estart[0]:Eend[-1]]
    x_star = x[Estart[0]:Eend[-1]]
    for i in range(len(Estart)):
        y_star[Estart[i]:Eend[i]+1] = 'E1'
        if i != len(Estart) - 1:
            y_star[Eend[i]+1:Estart[i+1]] = 'I'
    return ch

def code_sequence(augustus, sequences):
    df = pd.DataFrame(columns=['id','x','y'])
    
    for seq in sequences:
        s_id_aug = augustus.loc[augustus['chrom'] == seq.id]

        y = np.array(['N'] * s_id_aug.iloc[-1]['txEnd'],dtype='U')

        t = np.array([rec for rec in str(seq.seq)],dtype='U')[:s_id_aug.iloc[-1]['txEnd']]
        x = np.array([x.upper() if isinstance(x, str) else x for x in t])


        for index, row in s_id_aug.iterrows():

            if row['name'][-3:] == '.t2':
                y = code_Exon_Intron(row, x, y)
                
            check_t2 = row['name'][:-3] + '.t2'
            if s_id_aug[s_id_aug['name'].isin([check_t2])].empty:
                continue
            else: 
                y = code_Exon_Intron(row, y)
        # print(x[0],'\n',x[1],'\n/',x[2],'\n',x[3],'\n',x[100])
        print(len(x.tolist()))
        char = pd.Series({'id': seq.id, 'x':x.tolist(), 'y':y.tolist()})
        print(char['x'][:20])
        df = df.append(char,ignore_index=True)
    return df

In [7]:
temp = code_sequence(agsts, sequences)


23371726
['C', 'G', 'A', 'C', 'A', 'A', 'T', 'G', 'C', 'A', 'C', 'G', 'A', 'C', 'A', 'G', 'A', 'G', 'G', 'A']
25254583
['C', 'T', 'C', 'A', 'A', 'G', 'A', 'T', 'A', 'C', 'C', 'T', 'T', 'C', 'T', 'A', 'C', 'A', 'G', 'A']
27866355
['T', 'A', 'G', 'G', 'G', 'A', 'G', 'A', 'A', 'A', 'T', 'A', 'T', 'G', 'A', 'T', 'C', 'G', 'C', 'G']
32070586
['A', 'C', 'G', 'G', 'G', 'A', 'C', 'C', 'G', 'A', 'G', 'T', 'A', 'T', 'A', 'G', 'T', 'A', 'C', 'C']
1271726
['T', 'T', 'A', 'T', 'T', 'A', 'T', 'A', 'T', 'T', 'A', 'T', 'T', 'A', 'T', 'A', 'T', 'T', 'A', 'T']


In [9]:
def prepare_segment(x,y):
    df = pd.DataFrame(columns = ['x','y'])
    last_i = 0
    for i in range(1000,len(x),1000):
        df = df.append({'x': x[last_i:i], 'y':['START'] + y[last_i:i] + ['STOP']}, ignore_index = True)
        last_i = i
    return df

In [10]:
df = prepare_segment(temp.iloc[4]['x'],temp.iloc[4]['y'])

In [79]:
class MultiDimnDict():
    
    def __init__(self,axes):
        self.axes = axes

        self.data = {}
        prev = None

        for i in range(len(axes)-1, -1, -1):
            x = {}
            for n in axes[i]:
                if prev == None:
                    x[n] = 0
                else:
                    x[n] = copy.deepcopy(prev)
            prev = x
            self.data = x

    def __get_string(self, x,_str, depth):

        if type(list(x.values())[0]) == type(dict()):
            
            for i in x:

                _str += self.__get_string(x[i], depth * " " + i + "  \n", depth+1)
                # print(_str)
            return _str
        else:
            return _str + " " * depth +str(x) +"\n"

    def __str__(self):
        string = self.__get_string(self.data, "", 0)
        return string

    def add_value_helper(self,lst,val, temp, depth):
        label = lst[0]
        if len(lst) == 1:
            if label == ':':
                for x in self.axes[depth]:
                    temp[x] += val
            else:
                temp[label] += val
            return temp
        
        # print(lst[1:],val,temp[lst[0]])
        if label == ':':
            for x in self.axes[depth]:
                temp[x] = self.add_value_helper(lst[1:], val, temp[x], depth+1)
        else:
            temp[label] = self.add_value_helper(lst[1:],val,temp[label], depth+1)
        return temp


    def add_value(self,lst,val):
        if len(lst) == len(self.axes):
            self.data =  self.add_value_helper(lst,val,self.data, 0)
            return self
        else:
            return 0

    def to_tensor(self):
        key_x = {}
        key_x[i] = val for i,val in enumerate(key_x)
        key_y = {}
        key_y[i] = val for i,val in enumerate(key_y)

        
    
    

x= MultiDimnDict([['A', 'N', 'G', 'C', 'T'], ['E', 'START', 'STOP', 'N', 'I'], ['E', 'START', 'STOP', 'N', 'I']])

print(x.add_value(['A','N','N'], 1))


A  
 E  
  {'E': 0, 'START': 0, 'STOP': 0, 'N': 0, 'I': 0}
 START  
  {'E': 0, 'START': 0, 'STOP': 0, 'N': 0, 'I': 0}
 STOP  
  {'E': 0, 'START': 0, 'STOP': 0, 'N': 0, 'I': 0}
 N  
  {'E': 0, 'START': 0, 'STOP': 0, 'N': 1, 'I': 0}
 I  
  {'E': 0, 'START': 0, 'STOP': 0, 'N': 0, 'I': 0}
N  
 E  
  {'E': 0, 'START': 0, 'STOP': 0, 'N': 0, 'I': 0}
 START  
  {'E': 0, 'START': 0, 'STOP': 0, 'N': 0, 'I': 0}
 STOP  
  {'E': 0, 'START': 0, 'STOP': 0, 'N': 0, 'I': 0}
 N  
  {'E': 0, 'START': 0, 'STOP': 0, 'N': 0, 'I': 0}
 I  
  {'E': 0, 'START': 0, 'STOP': 0, 'N': 0, 'I': 0}
G  
 E  
  {'E': 0, 'START': 0, 'STOP': 0, 'N': 0, 'I': 0}
 START  
  {'E': 0, 'START': 0, 'STOP': 0, 'N': 0, 'I': 0}
 STOP  
  {'E': 0, 'START': 0, 'STOP': 0, 'N': 0, 'I': 0}
 N  
  {'E': 0, 'START': 0, 'STOP': 0, 'N': 0, 'I': 0}
 I  
  {'E': 0, 'START': 0, 'STOP': 0, 'N': 0, 'I': 0}
C  
 E  
  {'E': 0, 'START': 0, 'STOP': 0, 'N': 0, 'I': 0}
 START  
  {'E': 0, 'START': 0, 'STOP': 0, 'N': 0, 'I': 0}
 STOP  
  {'E': 0, 'STAR

Cant write to dataframe since the character array turns into shit

Steps

1. Create the count matrices
2. Use viterbi algo

In [93]:
class HMM():
    def __init__(self,data,num_x,num_y,vocab_x,vocab_y):
        self.data = data
        self.vocab_x = vocab_x
        self.vocab_y = vocab_y
        self.num_x = num_x
        self.num_y = num_y
    
    def helper(self, c1,c2,c3, vocab):
        if len(vocab) == 1:
            for i in vocab[0]:
                if c3[i] > 0:
                    c1[i] = c2[i] / c3[i]
                else:
                    c1[i] = 0
            return c1
        else:
            for i in vocab[0]:
                c1[i] = self.helper(c1[i], c2[i], c3, vocab[1:])
            return c1


    def calculate_emissions(self):

        axes_for_emissions = []
        for i in range(self.num_x):
            axes_for_emissions.append('x')
        for i in range(self.num_y):
            axes_for_emissions.append('y')
        vocab_for_emissions = []
        for axe in axes_for_emissions:
            if axe == 'y':
                vocab_for_emissions.append(list(self.vocab_y))
            elif axe == 'x':
                vocab_for_emissions.append(list(self.vocab_x))
        self.emissions = MultiDimnDict(vocab_for_emissions)
        count_emit = MultiDimnDict(vocab_for_emissions)
        count_yi = MultiDimnDict([list(self.vocab_y)])

        for i, row in self.data.iterrows():
            _iter = 1
            x = row['x']
            y = row['y']

            while _iter < len(y) - 2:
                _data = []
                list_y = []
                for i in range(self.num_y):
                    if (_iter - i) >= 0:
                        list_y.append(y[_iter - i])
                    else:
                        list_y.append(':')
                list_y = list_y[::-1] # so that the list is []....,y-2, y-1, y]
                list_x = []
                for i in range(self.num_x):
                    if (_iter - i - 1) >= 0:
                        list_x.append(x[_iter - i - 1])       
                    else:
                        list_x.append(':')
                list_x = list_x[::-1]
                _data = list_x + list_y
                count_emit.add_value(_data,1)
                count_yi.add_value([list_y[-1]],1)

                _iter += 1
            


        

            
        self.emissions.data = self.helper(self.emissions.data, count_emit.data, count_yi.data, vocab_for_emissions)
        return self.emissions




    def viterbi(self, test_x):
        ## Final Goal: Find PI(n,v) where n is the length of y without start and stop
        ## We wont use PI we will maximize negative log pi
        ## Step 1: Find PI(1,v)

        k = 1
        viterbi_y = ['START']
        PI=[]
        _pi = []
        while k < len(test_x)+1:
            list_x = []
            list_y = []
            for i in range(self.num_x):
                if (k - i - 1) >= 0:
                    list_x.append(test_x[k - i - 1])       
                else:
                    list_x.append(':')
            list_x = list_x[::-1]
            for i in range(self.num_y):
                if i == 0:
                    continue
                elif (k-i) >= 0:
                    list_y.append(viterbi_y[k-i])
                else:
                    list_y.append(':')
            list_y = list_y[::-1]

            _pi = []
            for pred_y_k in self.vocab_y:
                _data = list_x + list_y + [pred_y_k]
                a = self.emissions.get_sum(_data)
                nlog_a = sys.maxsize
                if a > pow(10,-7):
                    nlog_a = -1 * math.log(a)
                if len(PI) == 0:
                    _pi.append(nlog_a)
                else: 
                    
                    _pi.append(nlog_a + PI[-1][list(self.vocab_y).index(viterbi_y[-1])])

            min_pi = min(_pi)
            min_y = list(self.vocab_y)[_pi.index(min_pi)]

            viterbi_y.append(min_y)
            PI.append(_pi)

            k+= 1
        print(len(PI), 'Should be 1000')
        print(len(viterbi_y), 'Should be 1001 including Start tag')


        PI = PI[::-1]
        y_star = 'STOP'
        final_Y = []
        final_Y.append(y_star)
        for i in range(self.num_y -1):
            final_Y.append(viterbi_y[len(viterbi_y)-1-i])
            del PI[0]
        for _PI in PI:
            pred_pi = []
            index = PI.index(_PI)
            list_x = []
            list_y = []
            for i in range(self.num_x):
                if (index - i) >= 0:
                    list_x.append(test_x[index - i])       
                else:
                    list_x.append(':')
            list_x = list_x[::-1]
            for i in range(self.num_y-1):
                if (index-i) >= 0:
                    list_y.append(viterbi_y[index-i])
                else:
                    list_y.append(':')
            list_y = list_y[::-1]
            for y in self.vocab_y:
                _data = list_x + [y] + list_y 
                emission = sys.maxsize
                a = self.emissions.get_sum(_data)
                if a > pow(10,-7):
                    emission = math.log(a) * -1
                pred_pi.append(_PI[list(self.vocab_y).index(y)] + emission)


            min_pi = min(pred_pi)
            final_y = list(self.vocab_y)[pred_pi.index(min_pi)]


            final_Y.append(final_y)

        final_Y.append('START')
        final_Y = final_Y[::-1]
        viterbi_y += ['STOP']
        return final_Y, viterbi_y
            

        



In [94]:
hmm = HMM(df,3,6 ,set(temp.iloc[0]['x']), set(temp.iloc[0]['y'] + ['START','STOP']))


In [95]:
start = timer()
hmm.calculate_emissions()
stop = timer()
print(stop-start)

ZeroDivisionError: division by zero

In [26]:
start = timer()
hmm.calculate_transitions()
stop = timer()
print(stop-start)

STOP
STOP
STOP
STOP
STOP
1.5163649559999612


In [81]:
print(hmm.transitions)

E  
 {'E': 0.9876951821700254, 'STOP': 0.009955958894968638, 'START': 0.0, 'N': 0.00013345789403443215, 'I': 0.0022154010409715734}
STOP  
 {'E': 0, 'STOP': 0, 'START': 0, 'N': 0, 'I': 0}
START  
 {'E': 0.02940945191475977, 'STOP': 0.0, 'START': 0.0, 'N': 0.8570417551309271, 'I': 0.11354879295431312}
N  
 {'E': 4.586958177032733e-06, 'STOP': 0.009998651434295952, 'START': 0.0, 'N': 0.989996761607527, 'I': 0.0}
I  
 {'E': 0.0005687019724248898, 'STOP': 0.010021638416511776, 'START': 0.0, 'N': 0.0, 'I': 0.9894096596110633}



In [89]:
print(hmm.emissions)

A  
 E  
  {'E': 0, 'START': 0, 'STOP': 0, 'N': 0, 'I': 0}
 START  
  {'E': 0, 'START': 0, 'STOP': 0, 'N': 0, 'I': 0}
 STOP  
  {'E': 0, 'START': 0, 'STOP': 0, 'N': 0, 'I': 0}
 N  
  {'E': 0, 'START': 0, 'STOP': 0, 'N': 0, 'I': 0}
 I  
  {'E': 0, 'START': 0, 'STOP': 0, 'N': 0, 'I': 0}
N  
 E  
  {'E': 0, 'START': 0, 'STOP': 0, 'N': 0, 'I': 0}
 START  
  {'E': 0, 'START': 0, 'STOP': 0, 'N': 0, 'I': 0}
 STOP  
  {'E': 0, 'START': 0, 'STOP': 0, 'N': 0, 'I': 0}
 N  
  {'E': 0, 'START': 0, 'STOP': 0, 'N': 0, 'I': 0}
 I  
  {'E': 0, 'START': 0, 'STOP': 0, 'N': 0, 'I': 0}
G  
 E  
  {'E': 0, 'START': 0, 'STOP': 0, 'N': 0, 'I': 0}
 START  
  {'E': 0, 'START': 0, 'STOP': 0, 'N': 0, 'I': 0}
 STOP  
  {'E': 0, 'START': 0, 'STOP': 0, 'N': 0, 'I': 0}
 N  
  {'E': 0, 'START': 0, 'STOP': 0, 'N': 0, 'I': 0}
 I  
  {'E': 0, 'START': 0, 'STOP': 0, 'N': 0, 'I': 0}
C  
 E  
  {'E': 0, 'START': 0, 'STOP': 0, 'N': 0, 'I': 0}
 START  
  {'E': 0, 'START': 0, 'STOP': 0, 'N': 0, 'I': 0}
 STOP  
  {'E': 0, 'STAR

In [83]:
start = timer()
y_pred, vy = hmm.viterbi(temp.iloc[4]['x'])
stop = timer()
print(stop-start)


Done 1  [1.1577945174954207, 9223372036854775807, 9223372036854775807, 1.1354919087530715, 1.1043114825750735]
PI=  99
0     [3.8283668569803533, 9.223372036854776e+18, 9.223372036854776e+18, 4.080162186559459, 4.126571043154425]
1.424188367993338


In [84]:
print(len(y_pred),'\n',len(df.iloc[4]['y']))
print(y_pred[-1], df.iloc[4]['y'][-1])

102 
 102
STOP STOP


In [85]:
print(y_pred)

['START', 'I', 'E', 'I', 'E', 'I', 'I', 'I', 'I', 'I', 'I', 'I', 'I', 'I', 'I', 'I', 'I', 'E', 'E', 'I', 'I', 'I', 'I', 'E', 'E', 'I', 'I', 'I', 'E', 'E', 'I', 'E', 'E', 'I', 'I', 'E', 'E', 'E', 'I', 'E', 'E', 'I', 'I', 'I', 'I', 'E', 'E', 'E', 'E', 'E', 'I', 'I', 'I', 'I', 'I', 'I', 'I', 'I', 'I', 'E', 'E', 'I', 'I', 'I', 'E', 'I', 'I', 'I', 'E', 'E', 'E', 'E', 'E', 'I', 'I', 'E', 'E', 'E', 'I', 'I', 'E', 'E', 'E', 'E', 'I', 'E', 'I', 'E', 'E', 'E', 'E', 'E', 'E', 'E', 'E', 'I', 'E', 'E', 'E', 'E', 'E', 'STOP']


In [86]:
count = 0
y_corr = ['START'] + df.iloc[4]['y'] + ['STOP']
for i in range(len(y_pred)):
    if y_pred[i] == y_corr[i]:
        count += 1
print(count/len(y_pred))

0.00980392156862745


In [87]:

print(set(y_corr))
print(set(temp_y))
print(set(y_pred))
print([y_corr.count(i) for i in set(y_corr)])
print([temp_y.count(i) for i in set(temp_y)])
print([y_pred.count(i) for i in set(y_pred)])

{'N', 'STOP', 'START'}
{'I', 'E'}
{'I', 'STOP', 'E', 'START'}
[100, 2, 2]
[56, 43]
[51, 1, 49, 1]


In [90]:
start = timer()
y_pred,temp_y = hmm.viterbi(temp.iloc[4]['x'][:20000])
stop = timer()
print(stop-start)


Done 1  [1.1577945174954207, 9223372036854775807, 9223372036854775807, 1.1354919087530715, 1.1043114825750735]
PI=  19999
15000     [38168.282978398754, 9.223372036854815e+18, 9.223372036854815e+18, 38167.86864575531, 38167.80596789631]
10000     [25408.808828052264, 9.2233720368548e+18, 9.2233720368548e+18, 25408.31594031823, 25408.273514414625]
5000     [12661.542118926282, 9.223372036854788e+18, 9.223372036854788e+18, 12661.637173086183, 12661.646865710436]
0     [3.8283668569803533, 9.223372036854776e+18, 9.223372036854776e+18, 4.080162186559459, 4.126571043154425]
3.2705980879982235


In [95]:
print(len(y_pred),'\n',len(temp.iloc[4]['y'][:20000]))
print(y_pred[-1], temp.iloc[4]['y'][-1])

20002 
 20000
STOP N


In [98]:
count = 0
y_corr = ['START'] + temp.iloc[4]['y'][:20000] + ['STOP']
print(len(y_pred), len(y_corr))
for i in range(len(y_pred)):
    if y_pred[i] == y_corr[i]:
        count += 1
print(count/len(y_pred))

20002 20002
9.999000099990002e-05


In [99]:

print(set(y_corr))
print(set(temp_y))
print(set(y_pred))
print([y_corr.count(i) for i in set(y_corr)])
print([temp_y.count(i) for i in set(temp_y)])
print([y_pred.count(i) for i in set(y_pred)])

{'N', 'STOP', 'START'}
{'I', 'E'}
{'I', 'STOP', 'E', 'START'}
[20000, 1, 1]
[10028, 9971]
[10106, 1, 9894, 1]
